### DISCUSSION TITLE
[Chapter 12] @Order, @Async

### DISCUSSION THREAD

**Discussion by BvrPark** [cowaine/spring-boot-for-msa#199] [Chapter 12. Spring Events] [answered]
### Q1
p.733에 
> `@Async`가 public 메서드만 비동기로 동작한다.

라는 말이 있던데 이게 `@Async`가 스프링 AOP로 구현된 메커니즘이라 protected나 private은 프록시 객체에서 접근이 불가능해서 이것도 그런건지 궁금합니다.

### Q2
p.736 마지막 문단에
> 구독 메서드가 동작할 때 순서가 필요하다면 `@Order` 애너테이션을 사용한다.

라고 되어있는데 그럼 `@Order`의 옵션에서 우선 순위가 동일한 경우가 여러 개 존재하는 경우에는 어떻게 처리할까요...? 스프링 빈에 등록된 순서대로 처리할까요? 아니면 뭔가 다른 추가로 설정을 할 수 있는 방법이 존재할까요?

**Comment by ahngilwoong** [comment, marked as answer]
Q1. 말씀하신 부분이 맞는 것 같습니다.


Q2. 

[참고](https://unhosted.tistory.com/79)

위 사이트를 보면 같은 패키지 내라면 클래스 명 오름차순 순 패키지 외라면 패키지 명 오름차순 순으로 적용이 된다고 합니다.
더욱 자세히 설명하자면 빈이 먼저 선언된 순서대로 인데 
보통 패키지 내에서 클래스들이 알파벳 순으로 정렬되어 a부터 먼저 선언되었다고 판단하니 이름에대해 오름차순이라고 판단하는 것 같습니다.
정확히는 빈이 먼저 선언된 순서가 맞다 라고 생각합니다.

**Comment by JoisFe** [comment]
Q1)
- `@Async` 경우 두가지 제한이 있다고 합니다.
1. 오직 public method에만 사용 가능
2. Self-invocation : 동일한 클래스 내에서는 async method를 호출하는 자체 호출을 할 수 없음

그 이유로는 @BvrPark 님께서 생각하신대로 @Async를 사용할 때 비동기적인 메서드 호출을 프록시 객체를 통해 구현하고 해당 프록시 객체는 인터페이스 기반의 프록시를 생성하고, 프록시된 메서드 호출 시 실제 메서드를 비동기적으로 호출하는 방식으로 동작한다고 합니다. 이러한 방식이 AOP 기능을 이용하는 것이고 AOP 프록시는 대상 메서드가 인터페이스에 선언된 메서드에 해당해야 하므로 public 메서드에만 @Async를 사용할 수 있다고 합니다.

(`AOP 프록시는 대상 메서드가 인터페이스에 선언된 메서드에 해당해야 하므로` -> 결국 @BvrPark 님이 말씀하셨든 프록시 객체에서 접근이 가능하게 하려는 의도도 있다고 봅니다.)

- [참고자료](https://www.baeldung.com/spring-async#the-async-annotation)

Q2) 우선 순위가 동일하다 -> 순서가 중요하지 않다 라는 의미라면 따로 `@Order` 옵션을 둘 필요가 없다고 생각합니다. 만약 다른 우선순위가 있는 구독이 있다면 해당 우선순위가 같은 것들을 다른 우선순위가 있는 구독보다 우선순위보다 높게 혹은 낮은 번호로 동일하게 두면 되지 않을까 싶습니다.

스프링 빈의 경우 우선순위를 지정하지 않은 상태에서는 (order를 두지않으면_ 빈 초기화 순서를 보장하지 않는 정책이 있어 스프링 빈에 등록된 순서대로 처리함을 보장할 수 없습니다.

order 값이 중복되면 위 @ahngilwoong 님이 올려주신 부분을 참고하시면 될 것 같습니다.

**Reply by corock** [reply]
슬픈 현실이지만 실무에서 이런 일도 있긴 합니다.

1. A 가 `Order(1)` 코딩을 함
2. A 가 6개월 뒤 퇴사
3. 대체 인력으로 B 가 입사
4. B 가 동일한 코드에 신규로 추가하고 `Order(1)` 를 넣음
